### PR TITLE
[DF] Add test for Fill with object with arbitrary signature

### LINKED ISSUE #9428: [DF] Fill method does not support arbitrary signatures
According to the documentation of RDataFrame's [Fill method](https://root.cern/doc/master/classROOT_1_1RDF_1_1RInterface.html#ac1b7fe6b530c21f7917d55afe059ae08), the following should be supported:

```cpp
#include <ROOT/RDataFrame.hxx>
#include <TApplication.h>
#include <TH2D.h>

struct Jet {
  double a, b;
};

struct Boo {
  TH2D h{"", "", 10, 0, 10, 10, 0, 10};

  void Fill(const Jet &j) { h.Fill(j.a, j.b); }

  void Merge(const std::vector<Boo *> &others) {
    TList l;
    for (auto *o : others) {
      l.Add(&o->h);
      h.Merge(&l);
    }
  }

  void Draw() { h.Draw(); }
};

int main() {
  TApplication app("app", nullptr, nullptr);

  auto res = ROOT::RDataFrame(10)
               .Define("Jet", [] { return Jet{1., 2.}; })
               .Fill<Jet>(Boo{}, {"Jet"});
  res->Draw();

  app.Run();
}
```
Instead, `FillParHelper` only supports `Boo::Fill` signatures that take combinations of doubles and collections thereof.

https://github.com/root-project/root/pull/7499 goes in the right direction in generalizing `FillParHelper::Exec` but it still needs a fix to allow this general case.

### PR DESCRIPTION
**Pull request by eguiraud**
This is the last piece missing to fix #9428.

### CODE REVIEW

File: tree/dataframe/test/dataframe_interface.cxx [hunk 1]
```diff
@@ -776,3 +776,33 @@ TEST(RDataFrameInterface, SnapshotWithDuplicateColumns)
...
+   TH2D h{"", "", 10, 0, 10, 10, 0, 10};
+
+   void Fill(const Jet &j) { h.Fill(j.a, j.b); }
+
+   void Merge(const std::vector<Boo *> &others)
+   {
+      TList l;
+      for (auto *o : others) {
+         l.Add(&o->h);
+         h.Merge(&l);
```
Comment (Axel-Naumann) [inline tree/dataframe/test/dataframe_interface.cxx:794]: Doesn't this merge in the first iteration `others[0]`, in the second `others[1]` and again `others[0]`, etc? Should the test detect this?

File: tree/dataframe/test/dataframe_interface.cxx [hunk 2]
```diff
@@ -776,3 +776,33 @@ TEST(RDataFrameInterface, SnapshotWithDuplicateColumns)
...
+   double GetMeanX() const { return h.GetMean(1); }
+   double GetMeanY() const { return h.GetMean(2); }
+};
+
+// #9428
+TEST(RDataFrameInterface, FillCustomType)
+{
+   auto res = ROOT::RDataFrame(10).Define("Jet", [] { return Jet{1., 2.}; }).Fill<Jet>(Boo{}, {"Jet"});
+   EXPECT_DOUBLE_EQ(res->GetMeanX(), 1.);
+   EXPECT_DOUBLE_EQ(res->GetMeanY(), 2.);
```
Comment (Axel-Naumann) [inline tree/dataframe/test/dataframe_interface.cxx:807]: Consider checking the filled entries?

**Review by Axel-Naumann — APPROVED** [review]
But see comment.

File: tree/dataframe/test/dataframe_interface.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/test/dataframe_interface.cxx:794]: yikes, thanks. the Merge is only used for multi-thread runs and we are not testing that case here (we are not interested in the correctness of the helper, rather in checking that RDF can use helpers with arbitrary `Fill` signatures). I'll just leave `Merge` empty.

**Pull request merged**